terrain_simplification/include/terrain_simplification/terrain_simplification.hpp

```diff
@@ -0,0 +1,240 @@
+/**
+ * @file terrain_simplification.hpp
+ * @brief A module that simplifies the observed terrain, represented as a gridmap,
+ * by applying an OpenCV-based low-pass filter to extract the high-level geometry
+ * and qualify its irregularity by computing traversability using a filter chain.
+ * @author Oliwier Melon (omelon@robots.ox.ac.uk)
+ * @bug No known bugs.
+ * @date 07/01/2021
+ * @version 1.1
+ * @note The computation can be accelerated if the functionality of the Filter Chain
+ * were manually implemented.
+ * @copyright 2020, Oliwier Melon. BSD-3-Clause
+ */
+
+#ifndef TERRAIN_SIMPLIFICATION_HPP
+#define TERRAIN_SIMPLIFICATION_HPP
+
+#include <grid_map_core/grid_map_core.hpp>
+#include <grid_map_ros/grid_map_ros.hpp>
+#include <filters/filter_chain.h>
+
+#include <thread>
+#include <atomic>
+#include <boost/thread/pthread/shared_mutex.hpp>
+
+namespace terrain_simplification {
+
+class TerrainSimplification {
+
+public:
+  /**
+   * @brief Constructs a multi-threaded (2) object
+   */
+  TerrainSimplification();
+
+  /**
+   * @brief Destructs the multi-threaded (2) object
+   */
+  ~TerrainSimplification();
+
+protected:
+  /**
+   * @brief Assigns a shared pointer to the filter_chain_ object
+   * @param[in] filter_chain  shared pointer to a filter chain object in the derived class (ros).
+   */
+  void setFilterChain(
+      std::shared_ptr<filters::FilterChain<grid_map::GridMap> > filter_chain) {
+    filter_chain_ = filter_chain;
+  }
+
+  /**
+   * @brief Set the position and orientation of the robot
+   * @param[in] position     the position of the robot
+   * @param[in] orientation  the orientation of the robot
+   */
+  void setRobotPose (
+      const Eigen::Vector3d& position,
+      const Eigen::Quaterniond& orientation) {
+    robot_position_ = position;
```

Comment:
Instead of storing the position and the orientation separately, we could use an Eigen::Isometry3d